src/main/scala/chisel3/util/SparseVec.scala

```diff
@@ -0,0 +1,114 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3.util
+
+import chisel3._
+import chisel3.experimental.SourceInfo
+import chisel3.util.{log2Up, BitPat, Mux1H}
+import chisel3.util.experimental.decode.{decoder, TruthTable}
+import scala.collection.mutable.{LinkedHashMap, LinkedHashSet}
+
+sealed trait SparseVec[A] {
+
+  def apply(addr: UInt)(implicit _sourceInfo: SourceInfo): A
+
+}
+
+object SparseVec {
+
+  def binary[A <: Data](
+    size:    Int,
+    mapping: Seq[(Int, A)]
+  )(
+    implicit _sourceInfo: SourceInfo
+  ): SparseVec[A] = {
+    val addrWidth = log2Up(size)
+
+    val valueMap = LinkedHashMap.from(mapping)
+
+    val values = Wire(Vec(mapping.size + 1, chiselTypeOf(mapping.head._2)))
+
+    val valuesAddrWidth = log2Up(values.size)
+    values(0) := DontCare
+    val ttable = TruthTable(
+      valueMap.zipWithIndex.map {
+        case ((k, v), i) =>
+          values(i + 1) := v
+          BitPat(k.U(addrWidth.W)) -> BitPat((i + 1).U(valuesAddrWidth.W))
+      },
+      BitPat.N(addrWidth)
```

Comment:
Default to N? What if input address is not in the Sparse Vector? Should we assert it out?

Reply:
The default value should probably be configurable. This `SparseVec` is intended to replace this pattern for something like `sourceStall`: https://github.com/chipsalliance/rocket-chip/blob/f11fb1aea0033cf87aebcc9e5d3961810d1dcf57/src/main/scala/tilelink/ToAXI4.scala#L106 We've run into issues where this `Vec` gets comically large for certain parameterizations.

The test in this PR is modeled off of this example. Roughly:

1. You create a `Vec` `Wire`.
2. You set everything in the `Vec` to zero.
3. You set a few values in the `Vec`. This can be either constants (see: `sourceTable`) or expressions (see: `sourceStall`).
4. You dynamically index out of the `Vec`.

This is basically describing a decoder which _should be optimized by synthesis tools_. However, it looks really bad and can trip limits on dynamic indexes for "memories" for some tools. Specifically, Cadence Genus will warn if you try to dynamically index something larger that 16Ki.

There are some subtleties here where the dynamic index has both a default value and an out-of-bounds value. I'll put a comment later about this.

Reply:
The right thing to do here is probably return `BitPath.dontCare(addrWidth)` though this should be configurable.